staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -735,18 +638,38 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 	if !s.needFullSyncAndUnmark() {
 		// The set of nodes in the cluster hasn't changed, but we can retry
 		// updating any services that we failed to update last time around.
-		s.servicesToUpdate = s.updateLoadBalancerHosts(ctx, s.servicesToUpdate, workers)
+		// Besides, only latest Service specs should be applied.
+		var servicesToUpdate []*v1.Service
+		for key := range s.servicesToUpdate {
+			ns, svc, err := cache.SplitMetaNamespaceKey(key)
+			if err != nil {
+				klog.Errorf("failed to split meta namespace key %q: %v", key, err)
+				continue
+			}
+			cachedService, err := s.serviceLister.Services(ns).Get(svc)
+			if err != nil {
+				klog.Errorf("failed to get Service with key %q: %v", key, err)
+				continue
+			}
+			servicesToUpdate = append(servicesToUpdate, cachedService)
+		}
+
+		s.servicesToUpdate = s.updateLoadBalancerHosts(ctx, servicesToUpdate, workers)
 		return
 	}
 	klog.V(2).Infof("Syncing backends for all LB services.")
 
-	// Try updating all services, and save the ones that fail to try again next
+	// Try updating all services, and save the failed ones to try again next
 	// round.
-	s.servicesToUpdate = s.cache.allServices()
```

Comment:
allServices previously only listed all Services of Type=LoadBalancer, the new lister call would be listing all Services regardless of type.